File: backend/app/services/tokenizer.py
"""Word tokenizer shared by the reader engine.

Tokens are split on whitespace so trailing punctuation stays attached to its
word. The reader uses that punctuation to decide how long to hold each word.
"""

import re

_TOKEN_RE = re.compile(r"\S+")


def tokenize(text: str) -> list[str]:
    if not text:
        return []
    return _TOKEN_RE.findall(text.strip())


def count_words(text: str) -> int:
    return len(tokenize(text))

File: app/utils/tokenizer.ts
/**
 * Split text into display tokens. Whitespace-split keeps trailing punctuation
 * attached to its word so the reader can vary the hold time per token.
 * Mirrors the backend tokenizer so word indices line up across client/server.
 */
export function tokenize(text: string): string[] {
  if (!text) {
    return [];
  }
  return text.trim().split(/\s+/).filter(Boolean);
}

export function countWords(text: string): number {
  return tokenize(text).length;
}
